### PR TITLE
4500zenja1/Rewriting Question and Topic Controllers

### PR DESCRIPTION
**Pull request by 4500zenja1**
Also:
1. For Update commands the Id field must be non-nulled;
2. For GetList commands the Level field must be non-nulled, either.

Thanks to Andrei Rohau for showing me some of these changes.

### CODE REVIEW

File: Elevel/Application/Features/QuestionCommands/UpdateQuestionCommand.cs [hunk 1]
```diff
@@ -12,12 +12,7 @@ public class UpdateQuestionCommand
     {
         public class Request : IRequest<Response>
         {
-            public Guid Id { get; set; }
-            public string NameQuestion { get; set; }
-            public DateTimeOffset CreationDate { get; set; }
-            public bool Deleted { get; set; }
-            public Guid AnswerId { get; set; }
-            public Guid? AuditionId { get; set; }
+            public Guid? Id { get; set; }
```
Comment (andreyrogov8) [inline Elevel/Application/Features/QuestionCommands/UpdateQuestionCommand.cs:15]: Why did you delete all these fields?
What are going to update?

File: Elevel/Application/Features/TopicCommands/UpdateTopicCommand.cs [hunk 2]
```diff
@@ -15,7 +15,7 @@ public class UpdateTopicCommand
     {
         public class Request : UpsertTopicCommand.Request, IRequest<Response>
         {
-            public Guid Id { get; set; }
+            public Guid? Id { get; set; }
```
Comment (andreyrogov8) [inline Elevel/Application/Features/TopicCommands/UpdateTopicCommand.cs:18]: What are you going to update?

Comment (andreyrogov8) [reply Elevel/Application/Features/TopicCommands/UpdateTopicCommand.cs:18]: Why is this Guid should be nullable?

File: Elevel/Elevel.Api/Controllers/TopicController.cs [hunk 3]
```diff
@@ -17,41 +17,36 @@ public TopicController()
         }
 
         [HttpPost]
-        public async Task<IActionResult> Create([FromBody]CreateTopicCommand.Request request)
+        public async Task<IActionResult> Create([FromBody] CreateTopicCommand.Request request)
         {
             return Ok(await Mediator.Send(request));
```
Comment (andreyrogov8) [inline Elevel/Elevel.Api/Controllers/TopicController.cs:22]: What it should return if the Topic will not be created?
200 Ok?

File: Elevel/Elevel.Api/Controllers/TopicController.cs [hunk 4]
```diff
@@ -17,41 +17,36 @@ public TopicController()
...
+            var response = await Mediator.Send(request);
+            return response == null ? BadRequest() : Ok(response);
         }
 
-        [HttpGet("id:Guid")]
-        public async Task<IActionResult> GetById([FromRoute]GetTopicByIdQuery.Request request)
+        [HttpGet("{id:Guid}")]
+        public async Task<IActionResult> GetById([FromRoute] GetTopicByIdQuery.Request request)
         {
             return Ok(await Mediator.Send(request));
```
Comment (andreyrogov8) [inline Elevel/Elevel.Api/Controllers/TopicController.cs:35]: What it should return if the Topic will be empty?
200 Ok?

File: Elevel/Elevel.Api/Controllers/TopicController.cs [hunk 5]
```diff
@@ -17,41 +17,36 @@ public TopicController()
...
 
         [HttpDelete("{id:Guid}")]
-        public async Task<IActionResult> DeleteAsync(Guid id,[FromRoute] DeleteTopicCommand.Request request)
+        public async Task<IActionResult> DeleteAsync([FromRoute] DeleteTopicCommand.Request request)
         {
-            if(id != request.Id)
-            {
-                return BadRequest();
-            }
             return Ok(await Mediator.Send(request));
```
Comment (andreyrogov8) [inline Elevel/Elevel.Api/Controllers/TopicController.cs:41]: What it should return if the Topic will not be deleted?
200 Ok?

File: Elevel/Application/Features/TopicCommands/GetTopicListQuery.cs [hunk 6]
```diff
@@ -34,8 +34,11 @@ public Handler(IApplicationDbContext context, IMapper mapper)
 
             public async Task<Response> Handle(Request request, CancellationToken cancellationToken)
             {
-                var topic = _context.Topics.AsNoTracking()
-                    .OrderBy(x => x.Id);
+                var topic = _context.Topics.AsNoTracking();
+
+                if (request.Level != null) {
```
Comment (whitechapel5991) [inline Elevel/Application/Features/TopicCommands/GetTopicListQuery.cs:39]: instead request.Level != null use request.Level.HasValue

File: Elevel/Elevel.Api/Controllers/QuestionController.cs [hunk 7]
```diff
@@ -14,41 +14,37 @@ public QuestionController()
         }
 
         [HttpGet]
-        public async Task<IActionResult> GetQuestionList(GetQuestionListQuery.Request request)
+        public async Task<IActionResult> GetQuestionList([FromQuery]GetQuestionListQuery.Request request)
         {
-            return Ok(await Mediator.Send(request));
+            var response = await Mediator.Send(request);
+            return response == null ? BadRequest() : Ok(response);
```
Comment (whitechapel5991) [inline Elevel/Elevel.Api/Controllers/QuestionController.cs:20]: In which case response could be equal null?

File: Elevel/Elevel.Api/Controllers/QuestionController.cs [hunk 8]
```diff
@@ -14,41 +14,37 @@ public QuestionController()
...
         [HttpPost]
-        public async Task<IActionResult> CreateQuestionAsync(CreateQuestionCommand.Request request)
+        public async Task<IActionResult> CreateQuestionAsync([FromBody] CreateQuestionCommand.Request request)
         {
             return Ok(await Mediator.Send(request));
         }
 
         [HttpPut("{id:Guid}")]
-        public async Task<IActionResult> UpdateQuestionAsync(Guid id, UpdateQuestionCommand.Request request)
+        public async Task<IActionResult> UpdateQuestionAsync([FromRoute] Guid id, [FromBody] UpdateQuestionCommand.Request request)
```
Comment (whitechapel5991) [inline Elevel/Elevel.Api/Controllers/QuestionController.cs:38]: I would only leave the request parameter.

File: Elevel/Elevel.Api/Controllers/QuestionController.cs [hunk 9]
```diff
@@ -14,41 +14,37 @@ public QuestionController()
...
         {
-            if (id != request.Id)
-            {
-                return BadRequest();
-            }
-            return Ok(await Mediator.Send(request));
+            request.Id = id;
```
Comment (whitechapel5991) [inline Elevel/Elevel.Api/Controllers/QuestionController.cs:40]: When can this be useful?

File: Elevel/Elevel.Api/Controllers/TopicController.cs (hunk 3 above)
Comment (whitechapel5991) [reply Elevel/Elevel.Api/Controllers/TopicController.cs]: @andreyrogov8 In this case will throw exception. in this case need global exception handler. It is not in the scope of this task.

File: Elevel/Elevel.Api/Controllers/TopicController.cs [hunk 10]
```diff
@@ -17,41 +17,36 @@ public TopicController()
...
         [HttpPut("{id:Guid}")]
-        public async Task<IActionResult> Update(Guid id,[FromBody] UpdateTopicCommand.Request request)
+        public async Task<IActionResult> Update([FromRoute] Guid id, [FromBody] UpdateTopicCommand.Request request)
         {
-            if (id != request.Id)
-            {
-                return BadRequest();
-            }
-            return Ok(await Mediator.Send(request));
+            request.Id = id;
```
Comment (whitechapel5991) [inline Elevel/Elevel.Api/Controllers/TopicController.cs:50]: When can this be useful?

**Review by whitechapel5991 — CHANGES_REQUESTED** [review]

File: Elevel/Application/Features/QuestionCommands/UpdateQuestionCommand.cs (hunk 1 above)
Comment (4500zenja1) [reply Elevel/Application/Features/QuestionCommands/UpdateQuestionCommand.cs:15]: Undeleted

**Review by andreyrogov8 — APPROVED** [review]

File: Elevel/Elevel.Api/Controllers/TopicController.cs (hunk 4 above)
Comment (andreyrogov8) [reply Elevel/Elevel.Api/Controllers/TopicController.cs:35]: Are you going to return 200 Ok when it is null?

**Review by andreyrogov8 — CHANGES_REQUESTED** [review]

File: Elevel/Elevel.Api/Controllers/QuestionController.cs [hunk 11]
```diff
@@ -14,41 +14,37 @@ public QuestionController()
...
+            return response == null ? BadRequest() : Ok(response);
         }
 
         [HttpPost]
-        public async Task<IActionResult> CreateQuestionAsync(CreateQuestionCommand.Request request)
+        public async Task<IActionResult> CreateQuestionAsync([FromBody] CreateQuestionCommand.Request request)
         {
-            return Ok(await Mediator.Send(request));
+            var response = await Mediator.Send(request);
+            return response == null ? BadRequest() : Ok(response);
```
Comment (DShcharbakou) [inline Elevel/Elevel.Api/Controllers/QuestionController.cs:33]: just return Ok(response) because response will never be null.

File: Elevel/Elevel.Api/Controllers/TopicController.cs [hunk 12]
```diff
@@ -17,41 +17,37 @@ public TopicController()
         }
 
         [HttpPost]
-        public async Task<IActionResult> Create([FromBody]CreateTopicCommand.Request request)
+        public async Task<IActionResult> Create([FromBody] CreateTopicCommand.Request request)
         {
-            return Ok(await Mediator.Send(request));
+            var response = await Mediator.Send(request);
+            return response == null ? BadRequest() : Ok(response);
```
Comment (DShcharbakou) [inline Elevel/Elevel.Api/Controllers/TopicController.cs:23]: just return Ok(response) because response will never be null.

**Review by DShcharbakou — CHANGES_REQUESTED** [review]

File: Elevel/Application/Features/TopicCommands/UpdateTopicCommand.cs [hunk 13]
```diff
@@ -13,9 +11,12 @@ namespace Elevel.Application.Features.TopicCommands
 {
     public class UpdateTopicCommand
     {
-        public class Request : UpsertTopicCommand.Request, IRequest<Response>
+        public class Request : IRequest<Response>
         {
             public Guid Id { get; set; }
+            public string TopicName { get; set; }
+            public Level Level { get; set; }
+            public DateTimeOffset CreationDate { get; set; }
```
Comment (DShcharbakou) [inline Elevel/Application/Features/TopicCommands/UpdateTopicCommand.cs:19]: CreationDate Set by ORM in configuration files

**Review by DShcharbakou — CHANGES_REQUESTED** [review]

File: Elevel/Application/Features/TopicCommands/UpdateTopicCommand.cs (hunk 13 above)
Comment (4500zenja1) [reply Elevel/Application/Features/TopicCommands/UpdateTopicCommand.cs]: Deleted

**Review by DShcharbakou — DISMISSED** [review]

**Review by andreyrogov8 — APPROVED** [review]

**Review by DShcharbakou — APPROVED** [review]
